thevoid/connection.cpp

```diff
@@ -23,6 +23,66 @@
 #include "stream_p.hpp"
 #include "stockreplies_p.hpp"
 
+namespace {
+
+const long int USECS_IN_SEC = 1000 * 1000;
+const long int NSECS_IN_SEC = USECS_IN_SEC * 1000;
+
+struct timespec& operator+= (struct timespec& lhs, const struct timespec& rhs) {
+	lhs.tv_sec += rhs.tv_sec;
+	lhs.tv_nsec += rhs.tv_nsec;
+
+	if (lhs.tv_nsec >= NSECS_IN_SEC) {
+		lhs.tv_sec += 1;
+		lhs.tv_nsec -= NSECS_IN_SEC;
+	}
+
+	return lhs;
+}
+
+struct timespec operator+ (struct timespec lhs, const struct timespec& rhs) {
+	return lhs += rhs;
+}
+
+struct timespec& operator-= (struct timespec& lhs, const struct timespec& rhs) {
+	lhs.tv_sec -= rhs.tv_sec;
+	lhs.tv_nsec -= rhs.tv_nsec;
+
+	if (lhs.tv_nsec < 0) {
+		lhs.tv_sec -= 1;
+		lhs.tv_nsec += NSECS_IN_SEC;
+	}
+
+	if (lhs.tv_sec < 0) {
+		lhs = {0, 0};
+	}
+
+	return lhs;
+}
+
+struct timespec operator- (struct timespec lhs, const struct timespec& rhs) {
+	return lhs -= rhs;
+}
+
+long int timespec_to_usec(const struct timespec& t) {
+	return t.tv_sec * USECS_IN_SEC + t.tv_nsec / 1000;
+}
+
+struct timespec gettime_now() {
+	struct timespec now;
+
+#ifdef CLOCK_MONOTONIC_RAW
+	if (clock_gettime(CLOCK_MONOTONIC_RAW, &now) == 0) {
+		return now;
+	}
+#endif
+
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	return now;
+}
```

Comment:
Well, some time ago I read several papers about the topic, e.g.:
http://nadeausoftware.com/articles/2012/04/c_c_tip_how_measure_elapsed_real_time_benchmarking

And most of them suggested to use `CLOCK_MONOTONIC_RAW`.

There was a patch to OpenJDK:
http://mail.openjdk.java.net/pipermail/hotspot-dev/2013-January/008105.html
and corresponding Zookeeper issue:
https://issues.apache.org/jira/browse/ZOOKEEPER-1616

The problem you are talking about is that underlying clock source may run at non-constant speed (and it needs to be adjusted), that's why it's OK to measure time between frequent events.
But these are just words. What is "long term measurement"? How long the time span should be to consider it as a "long time measurement"?

If you really worry about `CLOCK_MONOTONIC_RAW` usage, we can go with just `CLOCK_MONOTONIC`.